### DISCUSSION TITLE
Filter with Ardalis/SmartEnums

### DISCUSSION THREAD

**Discussion by sh-sharifjonov** [alirezanet/Gridify#163] [Q&A]
I'm using Ardalis SmartEnum as a property on the entity. It uses string type for value (to store the code) and that is what get persisted in the database column. How can i filtering this property?

**Comment by alirezanet** [comment]
Hi @ShuhratSharifjonovEskhata, it depends ... how would you do it using Linq? Could you give me a Linq query without using Gridify so I can understand the problem better?

**Reply by sh-sharifjonov** [reply]
I have an object using the Ardalis.SmartEnum library.
```csharp
  public class GenderEnum : SmartEnum<GenderEnum, string>
  {
      public static readonly GenderEnum Female = new("FEMALE", "Женский");
      public static readonly GenderEnum Male = new("MALE", "Мужской");
      public static readonly GenderEnum Unspecified = new("UNSPECIFIED", "Не указан");
  
      public GenderEnum(string name, string value) : base(name, value)
      {
      }
  }
```
In the database, this object is stored only as a property name, and when selected, it is converted into an object.
```csharp
  builder.Property(x => x.Gender)
      .HasMaxLength(100)
      .HasConversion(
          to => to != null ? to.Name : null,
          from => GenderEnum.FromName(from, false));
```
With such a query 
```var passenger = dbContext.Passengers.Where(p => p.Gender == "MALE");```
LINQ automatically converts the field to string type
![image](https://github.com/alirezanet/Gridify/assets/156634706/3c0ff398-eb9e-4e1e-b50e-4561cc6512a1)

but if I want to use Gridify and add
```AddMap("Gender", p => p.Gender);```
i see
`Object of type 'System.String' cannot be converted to type 'GenderEnum'`

if I add such a mapper
`AddMap("Gender", p => Convert.ToString(p.Gender));`
then the query will be like this
`DbSet<Passenger>().Where(o => Convert.ToString((string)o.OrderStatus) == __Value_0)`

Is it possible to disable conversion to a type that has a Gender field? Or are there any other methods to fix this error?

**Reply by alirezanet** [reply]
Maybe this example helps:

```csharp

var query = new QueryBuilder<Passenger>()
				.AddMap("gender", q => q.Gender, c => GenderEnum.FromName(c)) // using Convertor to convert string to GenderEnum 
				.AddCondition("gender=MALE")
				.BuildFilteringExpression();
				
query.Dump();


class Passenger
{
	public GenderEnum Gender{ get; set; }
}


public class GenderEnum : SmartEnum<GenderEnum, string>
{
	public static readonly GenderEnum Female = new("FEMALE", "Женский");
	public static readonly GenderEnum Male = new("MALE", "Мужской");
	public static readonly GenderEnum Unspecified = new("UNSPECIFIED", "Не указан");

	public GenderEnum(string name, string value) : base(name, value)
	{
	}
}
```

If your Gender property is a GenderEnum, you can use a converter in your mappings to convert it to the target type, the same way you're doing it for EntityFramework, Since SmartEnum object is not a primitive type, automatic mapping generation doesn't work on this class.

**Reply by sh-sharifjonov** [reply]
If there is no suitable value in ENUM, then an exception is issued. Is it possible to make an empty list appear?

**Reply by alirezanet** [reply]
this is more a SmartEnum challenge than Gridify problem but I would do something like this:

```csharp
var query = new QueryBuilder<Passenger>()
		.AddMap("gender", q => q.Gender, c => GenderEnum.TryFromName(c, out var result) ? result : GenderEnum.None) 
		.AddCondition("gender=Unknows")
		.BuildFilteringExpression();

query.Dump();

;

class Passenger
{
	public GenderEnum Gender { get; set; }
}


public class GenderEnum : SmartEnum<GenderEnum, String>
{
	public static readonly GenderEnum Female = new("FEMALE", "Женский");
	public static readonly GenderEnum Male = new("MALE", "Мужской");
	public static readonly GenderEnum Unspecified = new("UNSPECIFIED", "Не указан");
	
	public static readonly GenderEnum None = new("None", "None");

	public GenderEnum(string name, string value) : base(name, value)
	{
	}
}
```
